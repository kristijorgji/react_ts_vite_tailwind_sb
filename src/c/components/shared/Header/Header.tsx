import ThemeToggle from '@/c/components/shared/ThemeToggle/ThemeToggle.tsx';

export const Header = () => {
    return (
        <header
            className={
                'bg-header-bg border-b-header-border fixed top-0 right-0 left-0 z-50 flex h-14 items-center justify-between border-b-1 px-4'
            }
        >
            {/* Left: Logo or Title */}
            <div className="text-lg font-semibold"></div>

            {/* Right */}
            <div className={'space-x-1'}>
                <ThemeToggle className={'font-semibold'} />
            </div>
        </header>
    );
};
