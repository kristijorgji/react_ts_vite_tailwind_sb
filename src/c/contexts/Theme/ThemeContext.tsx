import { createContext } from 'react';

import type { ThemeValue } from '@/c/types/core.ts';

export type ThemeContextValue = {
    theme: ThemeValue;
    toggleTheme: VoidFunction;
};

export const ThemeContext = createContext<ThemeContextValue | undefined>(undefined);
ThemeContext.displayName = 'ThemeContext';
