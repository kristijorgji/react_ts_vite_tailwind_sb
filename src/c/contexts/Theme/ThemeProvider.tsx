import React, { useEffect, useMemo, useState } from 'react';

import { ThemeContext } from '@/c/contexts/Theme/ThemeContext.tsx';

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const [theme, setTheme] = useState<'light' | 'dark'>(localStorage.getItem('theme') === 'dark' ? 'dark' : 'light');

    useEffect(() => {
        document.documentElement.classList.toggle('dark', theme === 'dark');
        localStorage.setItem('theme', theme);
    }, [theme]);

    const toggleTheme = () => setTheme((prev) => (prev === 'light' ? 'dark' : 'light'));
    const contextValue = useMemo(() => ({ theme, toggleTheme }), [theme]);

    return <ThemeContext value={contextValue}>{children}</ThemeContext>;
};
