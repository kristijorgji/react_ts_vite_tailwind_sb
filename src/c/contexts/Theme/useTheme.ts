import { use } from 'react';

import { ThemeContext, type ThemeContextValue } from '@/c/contexts/Theme/ThemeContext.tsx';

export default function useTheme(): ThemeContextValue {
    return use(ThemeContext)!;
}
