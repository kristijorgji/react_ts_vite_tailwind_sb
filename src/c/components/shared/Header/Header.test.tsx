import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import { ThemeProvider } from '@/c/contexts/Theme/ThemeProvider';

import { Header } from './Header';

describe('Header', () => {
    it('renders header correctly (snapshot)', () => {
        const { asFragment } = render(
            <ThemeProvider>
                <Header />
            </ThemeProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders header and toggles theme on button click', () => {
        render(
            <ThemeProvider>
                <Header />
            </ThemeProvider>
        );

        // Find the theme toggle button
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();

        // Check initial text
        expect(button.textContent?.toLowerCase()).toMatch(/dark|light/);

        // Simulate click
        fireEvent.click(button);

        // After clicking, text should toggle
        expect(button.textContent?.toLowerCase()).toMatch(/light|dark/); // could be either depending on initial state
    });
});
