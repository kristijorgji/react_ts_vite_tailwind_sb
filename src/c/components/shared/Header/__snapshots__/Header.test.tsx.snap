// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > renders header correctly (snapshot) 1`] = `
<DocumentFragment>
  <header
    class="bg-header-bg border-b-header-border fixed top-0 right-0 left-0 z-50 flex h-14 items-center justify-between border-b-1 px-4"
  >
    <div
      class="text-lg font-semibold"
    />
    <div
      class="space-x-1"
    >
      <button
        class="font-semibold"
        type="button"
      >
        ğŸŒ™ common:dark
      </button>
    </div>
  </header>
</DocumentFragment>
`;
